design/20190708.certificate-request-crd.md

```diff
@@ -144,6 +144,18 @@ type CertificateRequestSpec struct {
 	// implies that the 'signing' usage is set
 	// +optional
 	IsCA bool `json:"isCA,omitempty"`
+
+  // MaxPathLen requests that the pathLen in the requested certificate's
+  // basicConstraints be set by an issuer, which controls the maximum number of CA
+  // certificates which can appear in the chain issued by the requested
+  // certificate.  Not all issuers support MaxPathLen, and it will be left unset if
+  // unsupported. If a certificate's MaxPathLen is 0, then it cannot issue CA
+  // certificates; if MaxPathLen is n > 0, this certificate may issue a CA
+  // certificate, which in turn may issue a CA certificate only if n-1 is greater
+  // than zero. Most intermediate certificates will usually want to have a pathLen
+  // set to the smallest value possible. If omitted, no MaxPathLen will be
+  // requested for the issued certificate.
+  MaxPathLen *int32 `json:"isCA,omitempty"`
```

Comment:
I think adding this field is a mistake for a number of reasons.

1. There is limited support for `maxPathLen` in cert-manager Issuers, namely
only the CA and SelfSigned Issuers. This means adding more unexpected behaviour
for clients who request for one thing and have another returned. It then makes
sense that this field is actually a feature of _those Issuer types_, rather
than a feature of a cert-manager CertifcateRequest.

2. PathLen I believe to be fundamentally a CA responsibility and concern.
Conceptually, clients should have no say over what basic constraints they should
be issued, i.e. CA certificates or PathLen. The Issuer should be the one with
the keys to the castle. cert-manager historically has not had amazing controls
over what certificates can be issued by the CA Issuer, and although we now have
the Approval API as part of CertifcateRequests, a vanilla cert-manager install
approves everything.  Moving the pathlen constraint as a Issuer field will be
the first time a user can constrain whether a client can request intermediate
certificates or not.

3. I think we should not be afraid and should actively encourage having
multiple Issuer resources fronting the same CA. Issuers should be a profile
about what CA will be signing certificates and how they will do that. A
typical setup I can see is `my-issuer-leaf` and `my-issuer-int` who both
reference the same CA but have different PathLen settings. It is far easier to
restrict "Alice can only request from `my-issuer-int`, Bob can only request
from `my-issuer-leaf`" since policy is written based on _resource names and
objects_, rather than decoding the contents of the request itself.

4. Adding `isCA` as a field to CRs was a mistake IMO. For the same reasons
mentioned, this is a CA concern and shouldn't be up to the client for what they
request. Whilst an argument can be made that "we have already added `isCA`, why
not just other basic contraints?", we should make an effort to move onto to the
correct path for future API additions.

Reply:
Making Issuers more configurable and dynamic would be an improvement but is a _fundamentally_ different thing to this suggestion. I'd be in favour of that feature request - or of adding support for "certificate profiles" - but it's completely irrelevant to this in my view. Most of the presented bullet points in [this comment](https://github.com/jetstack/cert-manager/pull/4277#discussion_r679051745) are arguing for that feature, but that feature isn't what's being discussed. The feature would be valid to add _in addition to this new field_, not _instead of it_.

Certificate signing requests are the method by which a client asks an issuer to issue a certificate. Issuers are absolutely free to reject CSRs entirely, or ignore or change certain parts of the request in some cases. If a user is surprised by their CSR being rejected by an issuer, their priors are wrong on what CSRs are for.

CA certificates are still just certificates, and there is absolutely no difference between an X.509 certificate signing request for a CA certificate versus a leaf certificate used for server identification (it would in fact be perfectly valid to request a certificate which both a CA and is also able to be used for client/server identification).

Adding this field allows our CertificateRequest resource - our wrapper of X.509 certificate signing requests - to more easily and effectively model absolutely valid certificate requests. A side effect of that is that it will significantly improve the experience and security of bootstrapping and maintaining a private PKI using cert-manager for people who choose to use it.

The same idea is present [in Vault](https://www.vaultproject.io/api-docs/secret/pki#generate-intermediate), so there's prior art here.

> Conceptually, clients should have no say over what basic constraints they should be issued, i.e. CA certificates or PathLen.

This is completely at odds with CSRs; there is absolutely no such restriction in X.509. If issuing an intermediate certificate using a CSR, it would absolutely be expected that the client would request whatever basicConstraints it required to do its job; it's totally fine for an issuer to reject that request, but there's nothing anywhere which says that making the request is invalid.

Reply:
I agree with Josh on this, and I also think it aligns more closely with how the Kubernetes CSR API works. There are multiple `signerNames` defined for different 'CA profiles' which ultimately dictates what options are set on the certificates that the signer issues.

> Certificate signing requests are the method by which a client asks an issuer to issue a certificate

This is correct, _however_ note that the x509 CSR does *not* contain a field for a client to request a specific `maxPathLen`, which I think is a counter-point to your comment below `This is completely at odds with CSRs` - we are having to add this as a field on our own CertificateRequest resource solely because the x509 CSR standard *does not* specify how a user could request a specific `maxPathLen` when they create an x509 CSR.

Additionally, I think it's important we continue to consider how the upstream types/issuers/signers work when we are modifying our own CertificateRequest resource. I think adding this as a per-Issuer-type option makes a lot more sense, and if we _do_ want to support users overriding this themselves at the time of creating a CertificateRequest, having an annotation key such as `ca.issuers.cert-manager.io/requested-max-path-length` would more accurately indicate that this option is one supported by the _CA issuer_.

Given the initial use-case for `maxPathLen`, I'd be intrigued to hear if having a per-CertificateRequest option for this is truly desirable. This _really_ feels like it is a CA policy concern rather than the concern of someone consuming the eventual signed certificate.

> A side effect of that is that it will significantly improve the experience and security of bootstrapping and maintaining a private PKI using cert-manager for people who choose to use it.

I'm not really confident that this _does_ improve the experience and security of bootstrapping a private PKI vs having the option on the Issuer that is being requested from - why does allowing it per-CertificateRequest pose a significant benefit vs having it as "CA policy" option?

Reply:
and regarding @JoshVanL's point (4) above, I'd also agree there too... I think `isCA` would have been better as an issuer-specific option, which would also simplify policy enforcement as we could effectively just disallow/allow certain classes of users from requesting from _signers/issuers_ more broadly, rather than having to inspect the parameters such as `isCA` on each CR...

For a large number of issuer types, the field is actually totally non-functional.

Reply:
For context, original request: https://github.com/jetstack/cert-manager/issues/2820

Use case is almost literally every intermediate in private PKI, for defence in depth

Reply:
I'm in favour of adding this field, because it's one of 2 x509 basic constraints. (https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9).

In response to above 

> PathLen I believe to be fundamentally a CA responsibility and concern.

We should absolutely have an issuing template on Issuers specifying constraints.

> Adding isCA as a field to CRs was a mistake IMO

There are only 2 basic constraints, and we support 1 but not the other. If the solution to this is removing the isCA field then that is also fine, but as-is is not correct. We should add maxPathLen or remove isCA.

> however note that the x509 CSR does not contain a field for a client to request a specific maxPathLen

While everything in a CSR except the public key is optional and ignored, you can send any x509 extension in your request, which includes basic constraints, key usages, etc.

Not having this as a request option means you cannot request a pathlen 1 certificate from a CA Issuer that would usually issue pathlen 2 certificates. Whether this is a problem? 🤷

Reply:
> While everything in a CSR except the public key is optional and ignored, you can send any x509 extension in your request, which includes basic constraints, key usages, etc.

Whilst this is true, there isn't a standard extension/UID to actually denote these two being requested.

> Use case is almost literally every intermediate in private PKI, for defence in depth

I agree that we should support `maxPathLen` - but "every intermediate in private PKI" does not mean that having this as an option on the *Issuer* isn't suitable, right? Based on the original referenced feature request, there's nothing to suggest that having this as an option on the Issuer wouldn't be suitable too?

> There are only 2 basic constraints, and we support 1 but not the other. If the solution to this is removing the isCA field then that is also fine, but as-is is not correct. We should add maxPathLen or remove isCA.

Agreed :) although removing a field is a lot harder than adding it back, hence the apprehension around rushing to add yet another one here too without thinking carefully about what our strategy here is, and what is actually most useful for users and administrators alike (i.e. as the 'owner' or 'manager' of a CA persona, would I _want_ users to have to even think about this field when creating intermediates, or would I rather set this 'safely' for them?)

Reply:
>> PathLen I believe to be fundamentally a CA responsibility and concern.
> We should absolutely have an issuing template on Issuers specifying constraints.

If we can all agree that we should absolutely have this on the *Issuer* resource, would it be better for us to add it there *today*, and then see if more use-cases arise that actually require it to be done on a per CertificateRequest basis too then? Seems like there _is_ broad agreement that this should be configurable on e.g. the CA issuer's spec

Reply:
> > > PathLen I believe to be fundamentally a CA responsibility and concern.
> 
> > We should absolutely have an issuing template on Issuers specifying constraints.
> 
> If we can all agree that we should absolutely have this on the _Issuer_ resource, would it be better for us to add it there _today_, and then see if more use-cases arise that actually require it to be done on a per CertificateRequest basis too then? Seems like there _is_ broad agreement that this should be configurable on e.g. the CA issuer's spec

I would be keen for this approach. We should also add this to the SelfSigned Issuer as they commonly bootstrap CA Issuers.

Reply:
> we are having to add this as a field on our own CertificateRequest resource solely because the x509 CSR standard does not specify how a user could request a specific maxPathLen when they create an x509 CSR.

> Whilst this is true, there isn't a standard extension/UID to actually denote these two being requested (on a CSR) - if there was, then we would not need this field (nor the isCA field) at all.

This is not the case; CSRs can contain requests for arbitrary extensions, which can include the basicConstraints extension, which is where maxpathlen is set. Defined in [RFC 2985 section 5.4.2](https://datatracker.ietf.org/doc/html/rfc2985#page-17), and is supported [in Go](https://pkg.go.dev/crypto/x509#CertificateRequest) to some extent.

It seems that the reason this mechanism isn't used is because it'll require OID / ASN.1 stuff that nobody wants to do; much easier to have a custom bit of logic outside of the PEM-encoded CSR, and especially so when basicConstraints is one of the only extensions that can be relied upon because it's so widely supported. 

I guess that's why cert-manager has IsCA, why Go treats basicConstraints specially in `x509.Certificate` and why Vault has flags for this stuff too. Still, the _concept_ of requesting arbitrary things including basicConstraints is absolutely in PKCS#10.

> If we can all agree that we should absolutely have this on the Issuer resource

I'm not necc. agreed on that but as I said before, I think talking about issuers is a completely different thing and while I'd support something like that, I think this specific addition to this specific resource is valuable whether or not we do that and I'm going to focus on this PR.

>  why does allowing it per-CertificateRequest pose a significant benefit vs having it as "CA policy" option?

Because the client can request exactly what it needs to do its job. If my pod needs a CA that can issue other CAs, it can request a cert with maxPathLen=1. If my pod needs a CA which shouldn't ever issue other CAs, it can request a cert with maxPathLen=0.

An alternative tie-in with setting `maxPathLen` on issuers is that it enables clients to request less. If the issuer says "only maxPathLen <= 2" then a client who knows they need `maxPathLen = 0` can request that, and there shouldn't be any reason for the issuer to reject that. The issuer cannot know what maxPathLen is appropriate for the client.

It's much the same as with SANs; issuers could have a policy which only allows issuance on `*.example.com` but they should still allow clients to request just `foo.example.com` to reduce attack surface.

Reply:
> Because the client can request exactly what it needs to do its job. If my pod needs a CA that can issue other CAs, it can request a cert with maxPathLen=1. If my pod needs a CA which shouldn't ever issue other CAs, it can request a cert with maxPathLen=0.

But I don't think we have evidence (i.e. use-cases) to support this as being needed on a per-CertificateRequest basis - given the 'default' in the case you describe here (if the user did not explicitly set `maxPathLen`) is to permit any length at all, the referenced CA is effectively 'unprotected' (from the CA administrators point of view), as they are purely relying on a consumer of their CA to set `maxPathLen` appropriately. Conversely, I think it'd make more sense for the manager/owner of a CA to be able to configure whether `isCA` is permissible as well as what the `maxPathLen` of issued certificates is.

Having this power at the CA level means that these Pod consumers can request a CA certificate and know that it is properly configured. I still don't think your argument here justifies having a per-CertificateRequest option.

>  If my pod needs a CA which shouldn't ever issue other CAs, it can request a cert with maxPathLen=0.

Right, but now pods are having to work out a _lot_ of best practices here. Alternatively, we could have an issuer named `leaf-only` and another named `intermediates-only`, and the relevant pod could request from either Issuer accordingly to their needs (and policy can easily be applied over these issuer/signer names). This is far more akin to how `signerNames` work in Kubernetes, with the `apiserver-client` and `apiserver-serving` signerNames that are built into Kubernetes.

> An alternative tie-in with setting maxPathLen on issuers is that it enables clients to request less. If the issuer says "only maxPathLen <= 2" then a client who knows they need maxPathLen = 0 can request that, and there shouldn't be any reason for the issuer to reject that. The issuer cannot know what maxPathLen is appropriate for the client.

Absolutely, but in this case we're _now_ discussing setting this policy on a per-issuer basis, which I think based on your statement here demonstrates there *is* value in allowing this to be configured per-Issuer. "further scoping" (i.e. requesting a pathLen smaller than the issuer's configured maximum) could well be a valuable feature, however I think the questions/issues raised by myself and Josh also present counterpoints as to why that *may* not be desirable (albeit mostly just in the face of a lack of clear use-cases that are not otherwise served by having Issuer-specific options).

---

What I'm really advocating for here is that we take a careful approach to this and don't roll out a new configuration option on a resource that is implemented by _many_ external entities aside from just ourselves, as this 'muddies' our API surface and makes it less clear when an option/field is actually ignored or paid attention to (thus devaluating the re-world security benefits of the field in the first place).

Adding this to Issuers *does* unblock the use-cases described in #2820, and *does not* prevent us from extending CertificateRequests to permit configuring this at that level in future (and even moreso, there's nothing really stopping a CA today implementing/supporting this via the 'extensions'/OID fields as you've noted, if they feel it necessary).

Reply:
> Because the client can request exactly what it needs to do its job. If my pod needs a CA that can issue other CAs, it can request a cert with maxPathLen=1. If my pod needs a CA which shouldn't ever issue other CAs, it can request a cert with maxPathLen=0.

> It's much the same as with SANs; issuers could have a policy which only allows issuance on *.example.com but they should still allow clients to request just foo.example.com to reduce attack surface.

I disagree with this point completely, and is why I would like to push this idea of issuer profiles more. The isCA and pathLen constraint are fundamentally different beasts to the identity of the signed certificate holder. One declares how the certificate can be used for singing (a CA concern), the other defines the identity of the holder (a client concern). Defining separation of access to having these constraints via different Issuer resources is the most logical approach imo.

It is reasonable for an entity to request a certificate for its identity that is verified out of band (we use Kubernetes UserInfo), but another to ask for CA capabilities.

Reply:
My big concern is really that CertificateRequest is the most community-focused aspect of the whole project, and adding fields to it that are _mostly_ ignored is not (in my opinion) a good practice, as it dilutes the value/assurances given by the field in the first place when it is set by users.

Reply:
> given the 'default' in the case you describe here (if the user did not explicitly set maxPathLen) is to permit any length at all, the referenced CA is effectively 'unprotected' (from the CA administrators point of view), as they are purely relying on a consumer of their CA to set maxPathLen appropriately.

The "protection" here would be to set maxPathLen on the CA certificate that would be signing the CSR. Adding this field to CertificateRequest enables admins to do that.

> (thus devaluating the re-world security benefits of the field in the first place).

> CertificateRequest is the most community-focused aspect of the whole project, and adding fields to it that are mostly ignored is not (in my opinion) a good practice, as it dilutes the value/assurances given by the field in the first place when it is set by users.

These are not convincing arguments to me. If an issuer doesn't support this field, there's no benefit to this field, sure. But if an issuer _does_ support this field, this allows devs + admins to make good choices.

I think the community are OK with the concept that some issuers don't support everything - I think people probably understand why they can't request an intermediate from letsencrypt and they are at peace with that limitation.

Some people will be confused if they set MaxPathLen and some random external issuer doesn't support it. That is a shame. But this field will allow people who _do_ understand things to get better defence-in-depth.

The fact that some issuers won't support it doesn't change the benefits to those that do. Having this field will encourage issuers to support it, in fact!

> The isCA and pathLen constraint are fundamentally different beasts to the identity of the signed certificate holder...

I'd argue there is absolutely _nothing_ different between them conceptually. A certificate request is a request for a set of features to appear in a certificate. maxPathLen is a feature that can appear in a certificate. so is the subjectAltName extension.

Whether an issuer supports a given feature is not relevant to the mechanism which allows a given feature to be requested.